Merge worldcrafter

murapix opened the pull request:
Huge changes all around here:
* Recipe functionality is nearly all condensed into the base `InWorldRecipe` class
* An event system's been added in
* All output types are merged into a single `DynamicOutput` with per-recipe output type blacklists
  * As such, duplicate recipes such as `ItemFluidTransform` and `FluidFluidTransform` have been merged into single types
* Fluid recipes now can have a duration and may require multiple fluid source blocks
* Recipes for item burning, lightning striking a block, killing an entity while it's standing on specific blocks/items
* and a few more, more minor changes

Some things still left to do in future efforts:
* Refactor JEI support to be dynamically constructed based on recipe components
* Add JEI support for all recipe types
* Add various events to the recipe Event system, including those from the Issues list
* Updating the docs to reflect the new changes

Should fix issues:
#10 
#18 
#20 (maybe not completely)
#27 
#36 

Adds functionality to allow:
#19 
#22 
#34 
#40

murapix committed:
Giant base recipe and output merge and overhaul

murapix committed:
Add initial duration trackers, more merging craft functionality

murapix committed:
Clean up references and add invulnerability

murapix committed:
Fix fluid mixin target and output type selection

murapix committed:
Fix some JEI and recipe issues

murapix committed:
Fix burning recipes, minor refactoring

murapix committed:
Add entity killing recipes

Blame #thecodewarrior

murapix committed:
Finish up recipe events

murapix committed:
Re-add fluid consumption

MaxNeedsSnacks commented:
Hey, there it is, nice! Thanks for the pull request, I'll try to put together a full review soon so we can see if stuff needs changing, but from a cursory glance, I really liked the direction you took with this!

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/integration/jei/InteractioJEIPlugin.java:
```diff
@@ -43,12 +56,32 @@ public void registerCategories(IRecipeCategoryRegistration registration) {
+        registration.addRecipes(InWorldRecipeType.BLOCK_ANVIL.getRecipes(), BlockAnvilSmashingCategory.UID);
+    }
+    
+    public static void setOutputLists(IIngredients ingredients, DynamicOutput output) {
```
I kind of want to stop using streams here (and in other possibly performance critical sections) as they *do* produce a fair amount of overhead. Maybe later down the line in another release we can replace the stream-related calls (`map`, `fitler`, etc.) with utility methods that just iterate over the collection

murapix replied:
Streams here in JEI I wouldn't worry about - especially since these here are called once when the recipe is loaded in, and the data's stored for any future lookups. In other sections though, I've just had bad past experience with pre-optimization (my *own* pre-optimization), so I would rather do some performance profiling first, just to see if there's any actual issues.

MaxNeedsSnacks replied:
Ah that's fair! I was kind of worried these would be called every time the recipe display is initialised when scrolling through recipe pages, but this should (hopefully) be fine then!

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/integration/jei/categories/BlockAnvilSmashingCategory.java:
```diff
@@ -86,28 +88,14 @@ public IDrawable getIcon() {
     public void setIngredients(BlockAnvilSmashingRecipe recipe, IIngredients ingredients) {
         // display input block as item
-        ingredients.setInputs(VanillaTypes.ITEM, recipe.getInput().getMatchingBlocks()
+        ingredients.setInputs(VanillaTypes.ITEM, recipe.getBlockInput().getMatching()
```
Same thing as above goes for this (and the other recipe categories as well), but it's not exactly a high-priority issue so don't worry about it too much for now

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/BlockAnvilSmashingRecipe.java:
```diff
@@ -1,154 +1,83 @@
         this.damage = damage;
+        
+        // after the craft, damage the anvil
+        this.postCraft.add((hitPos, info) -> {
```
Would it be possible to just migrate this to an event players could specify in JSON rather than applying it by default? If so, we could just get rid of the damage parameter and make it part of the event's parameters instead, streamlining JSON structure a fair bit

murapix replied:
Ah yeah, we could move this into an event without too much trouble

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/ItemAnvilSmashingRecipe.java:
```diff
@@ -1,112 +1,81 @@
+        super(id, inputs, null, null, output, true, json);
+        
+        // damage anvil after craft, if it still exist then keep going
+        this.keepCraftingConditions.add((entities, info) -> {
```
Similarly to the block anvil damage event, I wanna see whether we can migrate this to an *optional* event specific to this recipe type. Optimally, we'd wanna merge them into one condition obviously, but that probably won't be possible here considering the different use cases and more importantly, parameter types

murapix replied:
I can try to merge these, but yeah, different parameter types makes that iffy

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/ItemFluidRecipe.java:
```diff
@@ -0,0 +1,105 @@
+        this.consumeFluid = consumeFluid;
+        
+        this.postCraft.add(Events.events.get(new ResourceLocation("particle"))::accept);
+        this.onCraftEnd.add((items, info) -> {
```
Same thing as with the anvil recipes, could we try to generify this into an event somehow so we can get rid of the `consumeFluid` double entirely?

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/ItemLightningRecipe.java:
```diff
@@ -1,109 +1,35 @@
-    public NonNullList<Ingredient> getIngredients() {
-        return NonNullList.of(Ingredient.EMPTY, inputs.stream().map(RecipeIngredient::getIngredient).toArray(Ingredient[]::new));
+        
+        this.postCraft.add((entities, info) -> sendParticle(ParticleTypes.END_ROD, info.getWorld(), Vec3.atCenterOf(info.getPos())));
```
This could probably be replaced with the pre-registered particle event?

murapix replied:
Oh right, this was one of the earlier ones, that should be switched over the pre-registered event

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/base/InWorldRecipe.java:
```diff
@@ -0,0 +1,270 @@
+    {
+        Arrays.stream(EventType.values()).filter(type -> json.has(type.jsonName)).forEach(type -> {
+            JsonArray array = GsonHelper.getAsJsonArray(json, type.jsonName);
+            Stream<ResourceLocation> locs = Streams.stream(array.iterator())
```
As with everything, we should look into whether de-streamifying this is a good idea or not (i.e. whether the overhead caused by streams is critical here or not)

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/base/InWorldRecipe.java:
```diff
@@ -0,0 +1,270 @@
+
+    @Override
+    public boolean isSpecial() {
+        return true;
```
We might be fine setting this to false since I remember @LatvianModder saying something about KubeJS skipping special recipes during (de)serialisation, which might break with people trying to create in-world recipes in KubeJS

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/duration/DurationManager.java:
```diff
@@ -0,0 +1,92 @@
+            S state = tracker.getState(pos);
+            R recipe = entry.getKey();
+            int duration = entry.getValue() + 1;
+            Interactio.LOGGER.warn("|><|" + duration + "|><|" + input + "|><|" + state);
```
This should be moved to debug or commented out for a production release

murapix replied:
Woops, will do

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/base/InWorldRecipeType.java:
```diff
@@ -1,37 +1,61 @@
-    public static final InWorldRecipeType<ItemFluidTransformRecipe> ITEM_FLUID_TRANSFORM = create("item_fluid_transform", ItemFluidTransformRecipe.SERIALIZER);
-    public static final InWorldRecipeType<FluidFluidTransformRecipe> FLUID_FLUID_TRANSFORM = create("fluid_fluid_transform", FluidFluidTransformRecipe.SERIALIZER);
+    public static final InWorldRecipeType<ItemFireRecipe> ITEM_BURN = create("item_burning", ItemFireRecipe.SERIALIZER);
+    public static final InWorldRecipeType<ItemFluidRecipe> ITEM_FLUID = create("item_fluid_transform", ItemFluidRecipe.SERIALIZER);
```
This can probably just be `item_fluid` or some other simpler key. Since this is a major release, I'm not really worried about breaking JSON syntax ^^

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/ingredient/EntityIngredient.java:
```diff
@@ -0,0 +1,207 @@
+        } else if (json.has("entity")) {
+            EntityType<?> entity = IEntrySerializer.ENTITY.read(json);
+            return new SingleEntityList(entity);
+//        } else if (json.has("tag")) {
```
What were the problems with this during implementation?

murapix replied:
I'm guessing you mean the tag section that's commented out? I just wasn't able to find any tag registry or anything to pull from for those.

MaxNeedsSnacks replied:
Ah that should be `EntityTypeTags` I think

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/util/CraftingInfo.java:
```diff
@@ -1,4 +1,23 @@
+    
+    public JsonObject getJson() { return json; }
+    
+    public abstract BlockPos getPos();
```
We could move this to `Vec3` to allow for additional detail in events that use position (`DefaultInfo` would then just return the vector at the center of the block it's targetting)

murapix replied:
Yeah, that shouldn't be a problem

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/util/IEntrySerializer.java:
```diff
@@ -74,4 +75,25 @@ public void write(FriendlyByteBuf buffer, Block content) {
+        @Override
+        public EntityType<?> read(JsonObject json) {
+            ResourceLocation id = new ResourceLocation(GsonHelper.getAsString(json, "entity"));
+            return Optional.ofNullable(ForgeRegistries.ENTITIES.getValue(id))
```
Might want to move this to `architectury` registries down the line

MaxNeedsSnacks reviewed (approved):
Looks good to me overall!

Before merging, I'd still like to see whether we can address some of the things I've noted along the way, and I want to do some overall code cleanup and futureproofing before the actual public release, but generally this *should* be good to go!

murapix committed:
Address PR comments and expand Event structure

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/Events.java:
```diff
@@ -26,29 +34,117 @@
+        events.put(new ResourceLocation("particle"), (inputs, info, json) -> sendParticle(info));
+        tickEvents.put(new ResourceLocation("particle"), (inputs, state, info, json) -> sendParticle(info));
+        
+        events.put(new ResourceLocation("consumeFluids"), (inputs, info, json) -> {
```
Very tiny nitpick: ResourceLocations in Minecraft should generally use snake_case rather than camelCase

MaxNeedsSnacks replied:
Actually turns out this is not a nitpick at all: ResLocs **only** accept lowercase letters, so this will in fact crash in production

murapix replied:
ugh, right, I keep forgetting that's a rule, stupid nonsense restrictions... (I know why it's a thing, still stupid though)

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/Events.java:
```diff
@@ -26,29 +34,117 @@
+            FluidIngredient fluidInput = info.getRecipe().getFluidInput();
+            if (fluidInput == null) return;
+
+            double consumeFluid = Utils.parseChance(json, "consumeFluid");
```
We can rename this to "chance" since we've already established what this is a chance for.

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/Events.java:
```diff
@@ -26,29 +34,117 @@
+            }
+        });
+        
+        events.put(new ResourceLocation("breakAnvil"), (inputs, info, json) -> {
```
Same as above, plus I think "damage_anvil" seems like a better fit here

MaxNeedsSnacks commented on src/main/java/ky/someone/mods/interactio/recipe/Events.java:
```diff
@@ -26,29 +34,117 @@
+        });
+        
+        events.put(new ResourceLocation("breakAnvil"), (inputs, info, json) -> {
+            double damage = Utils.getDouble(json, "damage");
```
Same as above, renaming to "chance" might look better if we end up going with "damage_anvil"

MaxNeedsSnacks reviewed (changes requested):
Going into nitpick territory now, so this should be pretty much good to go!

MaxNeedsSnacks commented:
Maybe as a general question to toss into the room: We should probably decide on a uniform format for JSON keys (either snake_case or camelCase), so what do you think would fit better?

I personally think camelCase works better for keys, while Minecraft itself obviously encourages snake_case for values

murapix commented:
I'm personally just more of a fan of camelCase, although that does kinda come with the territory of being a programmer these days

murapix committed:
Fix ResLoc casing and rename chance values

MaxNeedsSnacks commented:
I'll be moving this to last call so that if anyone (most likely you or me) still has any changes to suggest, they can do so until Tuesday ^^

MaxNeedsSnacks committed:
Final pre-merge refactor

- Remove unnecessary version code
- Fix parseItemIngredients ignoring "key" param
- Code cleanup
- Do not use unstable Guava API
- Extract damage_anvil as static method
- Fix javadoc
- Base reformat

MaxNeedsSnacks commented:
Just refactored and merged this! Just out of interest: Do you happen to have example json recipes I can put on the wiki while I'm re-reworking it later on?

murapix commented:
Not exactly example recipes, but here's all the recipes I made to test the changes.
[jeff.zip](https://github.com/MaxNeedsSnacks/Interactio/files/6273947/jeff.zip)
